doc/source/architecture.rst

```diff
@@ -0,0 +1,225 @@
+Mechanical app in-memory inside of a Python program. There is no additional
+running process associated with it. Mechanical's data model is directly
+available within Python, which means that the fully CRUD data model of the
+mechanical API can be used.
+
+
+
+Distributed systems
+===================
+
+This section contains a very basic explanation of *distributed systems*. It is
+not meant be exhaustive and rigorous, but it instead introduces just the topics
+necessary as simply as possible to help you understand the choices made by the
+designers of PyMechanical.
+
+A distributed system is a software system that uses a network to distribute
+software across physical machines. With a distributed system, the individual
+pieces of that system do not share an address space and therefore cannot call
+functions of each other directly. Instead, they communicate with each other by
+sending messages to each other. Examples include email, multiplayer games, web
+apps, and high-performance computing, among other things.
+
+Distributed systems have unique characteristics when compared to classical
+software systems that share an address space. For instance, in a distributed
+system, any call can fail because of a problem with the network, and the caller
+can not always know whether a call has failed. For this reason, features of
+interest to distributed system designers, such as fault tolerance, redundancy,
+and idempotency, are not emphasized by classical software systems.
+
+In a distributed system, any computer can theoretically talk to any other
+computer. However, it is useful to divide them conceptually into clients and
+servers. Clients send requests to servers. Servers are expected return a
+response to the client.
+
+There are protocols [#f5]_ that describe how information moves between
+computers. We begin with an explanation of remote procedure calls.
+
+Remote procedure calls
+----------------------
+
+When one computer invokes a procedure on another computer using a network, it
+is said to have done an RPC. Unlike a normal procedure call, it can be
+unreliable and orders of magnitude slower. There also needs to be a handshake
+so that both sides understand how to interpret the bytes that move on the wire
+between computers [#f6]_. This handshake can be negotiated byte-per-byte, but
+in practice, the internet protocols provide useful conventions and standards
+that have been tried and tested.
+
+While standards are useful, it is not practical for every developer to craft
+packets to send over the wire by hand. Around the most popular conventions of
+the internet protocols are libraries and tooling in popular programming
+languages. These libraries and tooling can be opinionated in how RPCs can and
+should be written. Some offer low-level method invocation facilities, like gRPC
+and zeroMQ. Others take a view on how RPCs should be done. Two such views are
+*Representational State Transfer* (REST) and *Remote Method Invocation* (RMI).
+
```

Comment:
However, libraries and tooling can be opinionated (? Have requirements about?) in how RPCs can and should be written. Some offer low-level method invocation facilities, like gRPC and zeroMQ. Others take a view on how RPCs should be done (such as?). Two such views are *Representational State Transfer* (REST) and *Remote Method Invocation* (RMI).